File: apps/eventsservice/internal/notifications/consumer.go
package notifications

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"strconv"

	"github.com/twmb/franz-go/pkg/kgo"
)

const (
	ConsumerGroup = "notifications-consumer"

	topicEntityChanges = "entity-changes"
	topicActivity      = "activity"
)

var errMissingField = errors.New("missing event field")

type Consumer struct {
	client *kgo.Client
	repo   Repository
}

func NewConsumer(client *kgo.Client, repo Repository) *Consumer {
	client.AddConsumeTopics(topicEntityChanges, topicActivity)
	return &Consumer{client: client, repo: repo}
}

func (c *Consumer) Run(ctx context.Context) {
	for {
		fetches := c.client.PollFetches(ctx)
		if err := fetches.Err0(); err != nil && errors.Is(err, context.Canceled) {
			return
		}
		if fetches.IsClientClosed() {
			return
		}
		fetches.EachError(func(topic string, partition int32, err error) {
			if errors.Is(err, context.Canceled) {
				return
			}
			slog.Warn("notifications consumer fetch failed", "topic", topic, "partition", partition, "error", err)
		})
		for _, record := range fetches.Records() {
			if err := c.processRecord(ctx, record); err != nil {
				slog.Warn("notifications consumer record failed", "topic", record.Topic, "partition", record.Partition, "offset", record.Offset, "error", err)
				break
			}
			if err := c.client.CommitRecords(ctx, record); err != nil {
				slog.Warn("notifications consumer commit failed", "topic", record.Topic, "partition", record.Partition, "offset", record.Offset, "error", err)
				break
			}
		}
	}
}

func (c *Consumer) processRecord(ctx context.Context, record *kgo.Record) (err error) {
	defer func() {
		if recovered := recover(); recovered != nil {
			slog.Error("notifications consumer recovered panic", "topic", record.Topic, "partition", record.Partition, "offset", record.Offset, "panic", recovered)
			err = fmt.Errorf("notifications consumer panic: %v", recovered)
		}
	}()
	return c.dispatch(ctx, record.Topic, record.Value)
}

func (c *Consumer) dispatch(ctx context.Context, topic string, value []byte) error {
	var payload map[string]json.RawMessage
	if err := json.Unmarshal(value, &payload); err != nil {
		return fmt.Errorf("decode event payload: %w", err)
	}

	switch topic {
	case topicEntityChanges:
		return c.dispatchEntityChange(ctx, payload)
	case topicActivity:
		return c.dispatchActivity(ctx, payload)
	default:
		slog.Warn("notifications consumer skipping unknown topic", "topic", topic)
		return nil
	}
}

func (c *Consumer) dispatchEntityChange(ctx context.Context, payload map[string]json.RawMessage) error {
	table, err := stringField(payload, "table")
	if err != nil {
		return err
	}
	if table != "posts" {
		return nil
	}

	op, err := stringField(payload, "op")
	if err != nil {
		return err
	}
	switch op {
	case "delete":
		postID, err := int64Field(payload, "id")
		if err != nil {
			return err
		}
		return c.repo.DeleteByEntity(ctx, strconv.FormatInt(postID, 10), []string{"like", "repost", "reply"})
	default:
		slog.Warn("notifications consumer skipping unknown entity op", "op", op, "table", table)
		return nil
	}
}

func (c *Consumer) dispatchActivity(ctx context.Context, payload map[string]json.RawMessage) error {
	op, err := stringField(payload, "op")
	if err != nil {
		return err
	}

	switch op {
	case "like":
		return c.insertPostNotification(ctx, payload, "like")
	case "unlike":
		return c.deletePostNotification(ctx, payload, "like")
	case "repost":
		return c.insertPostNotification(ctx, payload, "repost")
	case "unrepost":
		return c.deletePostNotification(ctx, payload, "repost")
	case "reply":
		return c.insertReplyNotification(ctx, payload)
	case "unreply":
		replyPostID, err := int64Field(payload, "reply_post_id")
		if err != nil {
			return err
		}
		return c.repo.DeleteByEntity(ctx, strconv.FormatInt(replyPostID, 10), []string{"reply"})
	case "follow":
		return c.insertFollowNotification(ctx, payload)
	case "unfollow":
		actorID, recipientID, err := actorRecipient(payload)
		if err != nil {
			return err
		}
		return c.repo.DeleteByActorAndType(ctx, actorID, recipientID, "follow", strconv.FormatInt(int64(actorID), 10))
	default:
		slog.Warn("notifications consumer skipping unknown activity op", "op", op)
		return nil
	}
}

func (c *Consumer) insertPostNotification(ctx context.Context, payload map[string]json.RawMessage, notifType string) error {
	outboxID, actorID, recipientID, err := insertFields(payload)
	if err != nil {
		return err
	}
	if actorID == recipientID {
		return nil
	}
	postID, err := int64Field(payload, "post_id")
	if err != nil {
		return err
	}
	return c.repo.Insert(ctx, outboxID, recipientID, actorID, notifType, strconv.FormatInt(postID, 10))
}

func (c *Consumer) deletePostNotification(ctx context.Context, payload map[string]json.RawMessage, notifType string) error {
	actorID, recipientID, err := actorRecipient(payload)
	if err != nil {
		return err
	}
	postID, err := int64Field(payload, "post_id")
	if err != nil {
		return err
	}
	return c.repo.DeleteByActorAndType(ctx, actorID, recipientID, notifType, strconv.FormatInt(postID, 10))
}

func (c *Consumer) insertReplyNotification(ctx context.Context, payload map[string]json.RawMessage) error {
	outboxID, actorID, recipientID, err := insertFields(payload)
	if err != nil {
		return err
	}
	if actorID == recipientID {
		return nil
	}
	replyPostID, err := int64Field(payload, "reply_post_id")
	if err != nil {
		return err
	}
	return c.repo.Insert(ctx, outboxID, recipientID, actorID, "reply", strconv.FormatInt(replyPostID, 10))
}

func (c *Consumer) insertFollowNotification(ctx context.Context, payload map[string]json.RawMessage) error {
	outboxID, actorID, recipientID, err := insertFields(payload)
	if err != nil {
		return err
	}
	if actorID == recipientID {
		return nil
	}
	return c.repo.Insert(ctx, outboxID, recipientID, actorID, "follow", strconv.FormatInt(int64(actorID), 10))
}

func insertFields(payload map[string]json.RawMessage) (int64, int32, int32, error) {
	outboxID, err := int64Field(payload, "_outbox_id")
	if err != nil {
		return 0, 0, 0, err
	}
	actorID, recipientID, err := actorRecipient(payload)
	if err != nil {
		return 0, 0, 0, err
	}
	return outboxID, actorID, recipientID, nil
}

func actorRecipient(payload map[string]json.RawMessage) (int32, int32, error) {
	actorID, err := int32Field(payload, "actor_id")
	if err != nil {
		return 0, 0, err
	}
	recipientID, err := int32Field(payload, "recipient_id")
	if err != nil {
		return 0, 0, err
	}
	return actorID, recipientID, nil
}

func stringField(payload map[string]json.RawMessage, name string) (string, error) {
	raw, ok := payload[name]
	if !ok {
		return "", fmt.Errorf("%w: %s", errMissingField, name)
	}
	var value string
	if err := json.Unmarshal(raw, &value); err != nil {
		return "", fmt.Errorf("decode %s: %w", name, err)
	}
	return value, nil
}

func int32Field(payload map[string]json.RawMessage, name string) (int32, error) {
	value, err := int64Field(payload, name)
	if err != nil {
		return 0, err
	}
	if value < -2147483648 || value > 2147483647 {
		return 0, fmt.Errorf("%s out of int32 range", name)
	}
	return int32(value), nil
}

func int64Field(payload map[string]json.RawMessage, name string) (int64, error) {
	raw, ok := payload[name]
	if !ok {
		return 0, fmt.Errorf("%w: %s", errMissingField, name)
	}
	var value json.Number
	if err := json.Unmarshal(raw, &value); err != nil {
		return 0, fmt.Errorf("decode %s: %w", name, err)
	}
	parsed, err := value.Int64()
	if err != nil {
		return 0, fmt.Errorf("decode %s: %w", name, err)
	}
	return parsed, nil
}

File: apps/flowservice/src/notifications/consumer.rs
use rdkafka::Message;
use rdkafka::consumer::Consumer as _;
use sqlx::PgPool as DatabasePool;
use std::sync::Arc;

use crate::consumer_health::ConsumerProgress;
use crate::notifications::db::NotificationDb;

// post_id/reply_post_id stay in the outbox payload contract but are unused
// here now that entity_id lookups key on the public id instead.
#[derive(serde::Deserialize)]
#[serde(tag = "op", rename_all = "snake_case")]
enum ActivityEvent {
    Like {
        #[serde(rename = "_outbox_id")]
        outbox_id: i64,
        actor_id: i32,
        recipient_id: i32,
        #[allow(dead_code)]
        post_id: i64,
        post_public_id: String,
    },
    Unlike {
        actor_id: i32,
        recipient_id: i32,
        #[allow(dead_code)]
        post_id: i64,
        post_public_id: String,
    },
    Repost {
        #[serde(rename = "_outbox_id")]
        outbox_id: i64,
        actor_id: i32,
        recipient_id: i32,
        #[allow(dead_code)]
        post_id: i64,
        post_public_id: String,
    },
    Unrepost {
        actor_id: i32,
        recipient_id: i32,
        #[allow(dead_code)]
        post_id: i64,
        post_public_id: String,
    },
    Reply {
        #[serde(rename = "_outbox_id")]
        outbox_id: i64,
        actor_id: i32,
        recipient_id: i32,
        #[allow(dead_code)]
        reply_post_id: i64,
        reply_post_public_id: String,
    },
    Unreply {
        #[allow(dead_code)]
        reply_post_id: i64,
        reply_post_public_id: String,
    },
    Follow {
        #[serde(rename = "_outbox_id")]
        outbox_id: i64,
        actor_id: i32,
        recipient_id: i32,
    },
    Unfollow {
        actor_id: i32,
        recipient_id: i32,
    },
    #[serde(other)]
    Unknown,
}

#[derive(serde::Deserialize)]
struct EntityChangeEvent {
    table: String,
    op: String,
    // id/reply_post_ids stay in the payload contract but are unused here
    // now that cleanup keys on the public id instead.
    #[allow(dead_code)]
    id: Option<i64>,
    public_id: Option<String>,
    #[allow(dead_code)]
    reply_post_ids: Option<Vec<i64>>,
    reply_post_public_ids: Option<Vec<String>>,
}

pub async fn run(
    consumer: rdkafka::consumer::StreamConsumer,
    db: DatabasePool,
    progress: Arc<ConsumerProgress>,
    mut shutdown_rx: tokio::sync::watch::Receiver<bool>,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    consumer.subscribe(&["activity", "entity-changes"])?;

    // Mark the initial watch value as seen so changed() only fires on a real transition.
    shutdown_rx.borrow_and_update();

    loop {
        tokio::select! {
            _ = shutdown_rx.changed() => break,
            result = consumer.recv() => match result {
                Err(e) => tracing::warn!(error = %e, "broker recv error"),
                Ok(msg) => {
                    let payload = msg.payload().unwrap_or_default();
                    if let Err(e) = dispatch(&db, msg.topic(), payload).await {
                        tracing::warn!(
                            topic = msg.topic(),
                            offset = msg.offset(),
                            error = %e,
                            "notification event failed"
                        );
                        // Continue rather than break — transient errors should not
                        // permanently kill the consumer loop.
                        continue;
                    }
                    consumer
                        .commit_message(&msg, rdkafka::consumer::CommitMode::Async)
                        .ok();
                    progress.mark_committed(msg.partition(), msg.offset());
                }
            }
        }
    }

    Ok(())
}

async fn dispatch(
    db: &DatabasePool,
    topic: &str,
    payload: &[u8],
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    match topic {
        "entity-changes" => dispatch_entity_change(db, payload).await,
        "activity" => dispatch_activity(db, payload).await,
        other => {
            tracing::warn!(
                topic = other,
                "notifications consumer skipping unknown topic"
            );
            Ok(())
        }
    }
}

async fn dispatch_entity_change(
    db: &DatabasePool,
    payload: &[u8],
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let event: EntityChangeEvent = match serde_json::from_slice(payload) {
        Ok(e) => e,
        Err(e) => {
            tracing::warn!(error = %e, "notifications entity-change event has invalid json");
            return Ok(());
        }
    };

    if event.table != "posts" {
        return Ok(());
    }

    if event.op == "delete" {
        let public_id = match event.public_id {
            Some(id) => id,
            None => {
                tracing::warn!("notifications entity-change delete event missing public_id");
                return Ok(());
            }
        };
        db.delete_by_entity(&public_id, &["like", "repost"]).await?;
        // Reply notifications are keyed by the reply's own entity ID; the
        // parent's FK SET NULL orphans them instead of deleting them.
        for reply_public_id in event.reply_post_public_ids.unwrap_or_default() {
            db.delete_by_entity(&reply_public_id, &["reply"]).await?;
        }
    } else {
        tracing::warn!(
            op = %event.op,
            table = %event.table,
            "notifications consumer skipping unknown entity op"
        );
    }

    Ok(())
}

async fn dispatch_activity(
    db: &DatabasePool,
    payload: &[u8],
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let event: ActivityEvent = match serde_json::from_slice(payload) {
        Ok(e) => e,
        Err(e) => {
            tracing::warn!(error = %e, "notifications activity event has invalid json");
            return Ok(());
        }
    };

    match event {
        ActivityEvent::Like {
            outbox_id,
            actor_id,
            recipient_id,
            post_id: _,
            post_public_id,
        } => {
            if actor_id != recipient_id {
                db.insert(outbox_id, recipient_id, actor_id, "like", &post_public_id)
                    .await?;
            }
        }
        ActivityEvent::Unlike {
            actor_id,
            recipient_id,
            post_id: _,
            post_public_id,
        } => {
            db.delete_by_actor_and_type(actor_id, recipient_id, "like", &post_public_id)
                .await?;
        }
        ActivityEvent::Repost {
            outbox_id,
            actor_id,
            recipient_id,
            post_id: _,
            post_public_id,
        } => {
            if actor_id != recipient_id {
                db.insert(outbox_id, recipient_id, actor_id, "repost", &post_public_id)
                    .await?;
            }
        }
        ActivityEvent::Unrepost {
            actor_id,
            recipient_id,
            post_id: _,
            post_public_id,
        } => {
            db.delete_by_actor_and_type(actor_id, recipient_id, "repost", &post_public_id)
                .await?;
        }
        ActivityEvent::Reply {
            outbox_id,
            actor_id,
            recipient_id,
            reply_post_id: _,
            reply_post_public_id,
        } => {
            if actor_id != recipient_id {
                db.insert(
                    outbox_id,
                    recipient_id,
                    actor_id,
                    "reply",
                    &reply_post_public_id,
                )
                .await?;
            }
        }
        ActivityEvent::Unreply {
            reply_post_id: _,
            reply_post_public_id,
        } => {
            db.delete_by_entity(&reply_post_public_id, &["reply"])
                .await?;
        }
        ActivityEvent::Follow {
            outbox_id,
            actor_id,
            recipient_id,
        } => {
            if actor_id != recipient_id {
                db.insert(
                    outbox_id,
                    recipient_id,
                    actor_id,
                    "follow",
                    &actor_id.to_string(),
                )
                .await?;
            }
        }
        ActivityEvent::Unfollow {
            actor_id,
            recipient_id,
        } => {
            db.delete_by_actor_and_type(actor_id, recipient_id, "follow", &actor_id.to_string())
                .await?;
        }
        ActivityEvent::Unknown => {
            tracing::warn!("unknown activity event, skipping");
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn entity_change_event_deserializes_without_reply_post_ids() {
        let event: EntityChangeEvent =
            serde_json::from_str(r#"{"table":"posts","op":"delete","id":42}"#).unwrap();
        assert_eq!(event.id, Some(42));
        assert!(event.reply_post_ids.is_none());
        assert!(event.public_id.is_none());
        assert!(event.reply_post_public_ids.is_none());
    }

    #[test]
    fn entity_change_event_deserializes_with_reply_post_ids() {
        let event: EntityChangeEvent = serde_json::from_str(
            r#"{"table":"posts","op":"delete","id":42,"public_id":"11111111-1111-1111-1111-111111111111","reply_post_ids":[7,8],"reply_post_public_ids":["22222222-2222-2222-2222-222222222222","33333333-3333-3333-3333-333333333333"]}"#,
        )
        .unwrap();
        assert_eq!(event.reply_post_ids, Some(vec![7, 8]));
        assert_eq!(
            event.public_id,
            Some("11111111-1111-1111-1111-111111111111".to_string())
        );
        assert_eq!(
            event.reply_post_public_ids,
            Some(vec![
                "22222222-2222-2222-2222-222222222222".to_string(),
                "33333333-3333-3333-3333-333333333333".to_string(),
            ])
        );
    }
}
